pil-analyzer/src/expression_processor.rs

```diff
@@ -121,54 +114,60 @@ impl<'a, D: AnalysisDriver> ExpressionProcessor<'a, D> {
                     op,
                     right: r,
                 },
-            ) => Expression::BinaryOperation(
+            ) => Ok(Expression::BinaryOperation(
                 src,
                 BinaryOperation {
-                    left: Box::new(self.process_expression(*l)),
+                    left: Box::new(self.process_expression(*l)?),
                     op,
-                    right: Box::new(self.process_expression(*r)),
+                    right: Box::new(self.process_expression(*r)?),
                 },
-            ),
+            )),
             PExpression::UnaryOperation(src, UnaryOperation { op, expr: value }) => {
-                Expression::UnaryOperation(
+                Ok(Expression::UnaryOperation(
                     src,
                     UnaryOperation {
                         op,
-                        expr: Box::new(self.process_expression(*value)),
+                        expr: Box::new(self.process_expression(*value)?),
                     },
-                )
+                ))
             }
-            PExpression::IndexAccess(src, index_access) => Expression::IndexAccess(
+            PExpression::IndexAccess(src, index_access) => Ok(Expression::IndexAccess(
                 src,
                 parsed::IndexAccess {
-                    array: Box::new(self.process_expression(*index_access.array)),
-                    index: Box::new(self.process_expression(*index_access.index)),
+                    array: Box::new(self.process_expression(*index_access.array)?),
+                    index: Box::new(self.process_expression(*index_access.index)?),
                 },
-            ),
-            PExpression::FunctionCall(src, c) => Expression::FunctionCall(
+            )),
+            PExpression::FunctionCall(src, c) => Ok(Expression::FunctionCall(
                 src,
                 parsed::FunctionCall {
-                    function: Box::new(self.process_expression(*c.function)),
-                    arguments: self.process_expressions(c.arguments),
+                    function: Box::new(self.process_expression(*c.function)?),
+                    arguments: self.process_expressions(c.arguments)?,
                 },
-            ),
+            )),
             PExpression::MatchExpression(src, MatchExpression { scrutinee, arms }) => {
-                Expression::MatchExpression(
+                let scrutinee = self.process_expression(*scrutinee)?;
+                let arms = arms
+                    .into_iter()
+                    .map(|MatchArm { pattern, value }| {
+                        let vars = self.save_local_variables();
+                        let processed_pattern = self.process_pattern(pattern)?;
+                        let processed_value = self.process_expression(value)?;
+                        self.reset_local_variables(vars);
```

Comment:
I always try to keep the state correct even in case of an error. If we have an error here, the local variables still contain the state they have inside the match arm. What you could do is somethign like
```
let processed_pattern = self.process_pattern(pattern);
let processed_value = ...;
self.reset_local_variables(vars);
let processed_pattern = processed_pattern?;
```
(or use `?` when you construct the MatchArm below.